### PR TITLE
Add support for pointers to reflect.Kind types

### PR DESCRIPTION
**Pull request by pandada8**
this should resolve GODRIVER-535, pointers handling is added to both fast path and reflection path.

### CODE REVIEW

File: bson/bsoncodec/default_value_decoders.go [hunk 1]
```diff
@@ -62,33 +62,66 @@ func (dvd DefaultValueDecoders) RegisterDefaultDecoders(rb *RegistryBuilder) {
...
 	}
 
 	b, err := vr.ReadBoolean()
-	val.SetBool(b)
+	if err != nil {
+		return err
+	}
+
+	if val.Kind() == reflect.Ptr {
+		val.Set(reflect.New(val.Type().Elem()))
```
Comment (skriptble) [inline bson/bsoncodec/default_value_decoders.go]: We should only allocate a new value here if the val is `nil`.

File: bson/bsoncodec/default_value_decoders.go [hunk 2]
```diff
@@ -62,33 +62,66 @@ func (dvd DefaultValueDecoders) RegisterDefaultDecoders(rb *RegistryBuilder) {
...
 
 	val := reflect.ValueOf(i)
 	if !val.IsValid() || val.Kind() != reflect.Ptr || !val.Elem().CanSet() {
 		return errors.New("BooleanDecodeValue can only be used to decode settable (non-nil) values")
 	}
 	val = val.Elem()
-	if val.Type().Kind() != reflect.Bool {
+
+	if notKindAndPtrToKind(val, reflect.Bool) {
 		return ValueDecoderError{Name: "BooleanDecodeValue", Types: []interface{}{bool(true)}, Received: i}
```
Comment (skriptble) [inline bson/bsoncodec/default_value_decoders.go]: Types should be `*bool` and `**bool`.

File: bson/bsoncodec/default_value_decoders.go [hunk 3]
```diff
@@ -197,7 +277,12 @@ func (dvd DefaultValueDecoders) IntDecodeValue(dc DecodeContext, vr bsonrw.Value
 		}
 	}
 
-	val.SetInt(i64)
+	if val.Type().Kind() == reflect.Ptr {
+		val.Set(reflect.New(val.Type().Elem()))
```
Comment (skriptble) [inline bson/bsoncodec/default_value_decoders.go]: We should also only alloc if val is `nil`.

File: bson/bsoncodec/default_value_decoders.go [hunk 4]
```diff
@@ -452,17 +627,22 @@ func (dvd DefaultValueDecoders) Decimal128DecodeValue(dctx DecodeContext, vr bso
 		return fmt.Errorf("cannot decode %v into a decimal.Decimal128", vr.Type())
 	}
 
-	target, ok := i.(*decimal.Decimal128)
-	if !ok || target == nil {
+	switch target := i.(type) {
+	case *decimal.Decimal128:
+		d128, err := vr.ReadDecimal128()
```
Comment (skriptble) [inline bson/bsoncodec/default_value_decoders.go]: Is there a reason you moved the `ReadDecimal128` method invocation here? It adds repetitive code and extra errors to check.

File: bson/bsoncodec/struct_codec.go [hunk 5]
```diff
@@ -97,7 +97,13 @@ func (sc *StructCodec) EncodeValue(r EncodeContext, vw bsonrw.ValueWriter, i int
 		}
 
 		ectx := EncodeContext{Registry: r.Registry, MinSize: desc.minSize}
-		err = encoder.EncodeValue(ectx, vw2, rv.Interface())
+
+		// HACK: pass *Document as is
```
Comment (skriptble) [inline bson/bsoncodec/struct_codec.go]: Why is this here? The `bson.Document` decoder should only be handling `**bson.Document`.

**Comment by pandada8** [comment]
still finding the elegant checking process for most typing. I may push a new version in weekend

File: bson/bsoncodec/struct_codec.go [hunk 6]
```diff
@@ -86,8 +86,7 @@ func (sc *StructCodec) EncodeValue(r EncodeContext, vw bsonrw.ValueWriter, i int
 		if iz, ok := encoder.(CodecZeroer); ok {
 			iszero = iz.IsTypeZero
 		}
-
-		if desc.omitEmpty && iszero(rv.Interface()) {
+		if (rv.Kind() == reflect.Ptr || desc.omitEmpty) && iszero(rv.Interface()) {
```
Comment (delioda79) [inline bson/bsoncodec/struct_codec.go]: Hi, I was testing this but I think this line is wrong. If we do not have omitempty and the kind is pointer we want to add the key just with a null value in case the pointer is a nil pointer. Problem here is that iszero does not work with pointers and apparently is defined only for objectid

File: bson/bsoncodec/default_value_decoders.go (hunk 2 above)
Comment (pandada8) [reply bson/bsoncodec/default_value_decoders.go]: all related types should be updated

File: bson/bsoncodec/default_value_decoders.go (hunk 1 above)
Comment (pandada8) [reply bson/bsoncodec/default_value_decoders.go]: a new function `getValue` is introduced to check and allocate new value.  
do you have any idea of a better name ?

File: bson/bsoncodec/default_value_decoders.go (hunk 4 above)
Comment (pandada8) [reply bson/bsoncodec/default_value_decoders.go]: changed EmptyInterfaceDecodeValue's closure style

File: bson/bsoncodec/struct_codec.go (hunk 6 above)
Comment (skriptble) [reply bson/bsoncodec/struct_codec.go]: That is correct, we want to write a BSON Null if the pointer is nil, not skip it entirely.

File: bson/bsoncodec/default_value_decoders.go [hunk 7]
```diff
@@ -62,15 +63,88 @@ func (dvd DefaultValueDecoders) RegisterDefaultDecoders(rb *RegistryBuilder) {
...
+		typeString = append(typeString, payload)
+	}
+	return fmt.Errorf("cannot decode %v into a %s", vr.Type(), strings.Join(typeString, ", "))
+}
+
 // BooleanDecodeValue is the ValueDecoderFunc for bool types.
 func (dvd DefaultValueDecoders) BooleanDecodeValue(dctx DecodeContext, vr bsonrw.ValueReader, i interface{}) error {
-	if vr.Type() != bsontype.Boolean {
-		return fmt.Errorf("cannot decode %v into a boolean", vr.Type())
+	var (
```
Comment (skriptble) [inline bson/bsoncodec/default_value_decoders.go]: Instead of using a multiline `var` block, can you please change this to individual `var` statements?

File: bson/bsoncodec/default_value_decoders.go [hunk 8]
```diff
@@ -62,15 +63,88 @@ func (dvd DefaultValueDecoders) RegisterDefaultDecoders(rb *RegistryBuilder) {
...
+		err    error
+		b      bool
+		isNull bool
+	)
+	readValue := func() {
```
Comment (skriptble) [inline bson/bsoncodec/default_value_decoders.go]: I think we're adding a little too much indirection here. Let's instead just split off the fast path decoders into separate methods, and call them from this function if the type matches. It's fine to duplicate this specific logic between the methods.

File: bson/bsoncodec/struct_codec.go (hunk 6 above)
Comment (delioda79) [reply bson/bsoncodec/struct_codec.go]: So yes this code here is wrong as it is skipping any nil pointer

Comment (pandada8) [reply bson/bsoncodec/struct_codec.go]: I update the code, now `WriteNull` is called when an nil pointer is encountered.
Should I put it in a standalone function ?

File: bson/bsoncodec/default_value_decoders.go (hunk 7 above)
Comment (pandada8) [reply bson/bsoncodec/default_value_decoders.go]: all var block in default_value_decoders is changed

File: bson/bsoncodec/default_value_decoders.go (hunk 8 above)
Comment (pandada8) [reply bson/bsoncodec/default_value_decoders.go]: I am not sure I understood correctly, so I only update the Boolean case. If you think it's ok, I can work on the rest functions

File: bson/bsoncodec/struct_codec.go (hunk 6 above)
Comment (skriptble) [reply bson/bsoncodec/struct_codec.go]: For the encode path, we should just let the individual encoders handle it, we don't need to have the struct codec care if it has a `nil` pointer. We should add `nil` pointer handling for all of the  encoders as well as the decoders. This will allow users to create an encoder that handles `nil` pointers differently, e.g. if a user wants to put some default data instead of BSON null.

File: bson/bsoncodec/default_value_decoders.go (hunk 8 above)
Comment (skriptble) [reply bson/bsoncodec/default_value_decoders.go]: I added a patch with an updated [`BooleanDecodeValue`](https://github.com/mongodb/mongo-go-driver/commit/dc6a94e7fc31b407ad13fbd3d2e11a6ff069c76d#diff-d7870a7e54c1393f5080c1fbf611666dR114), can you update your patch to follow this design? If you want to do some things to make the code more compact, feel free.

**Comment by Trial97** [comment]
Hello guys,

How it's going with this PR?

Thanks,
Trial97

**Comment by pandada8** [comment]
It's been a long and tough week for me, moved into a new apartment and had a fever, hope it's not too late to submit some updated patch.

I rebased my branch to resolve some merge conflict and add new error type. more update on Decoder function should come in tomorrow.

File: bson/bsoncodec/struct_codec.go (hunk 6 above)
Comment (pandada8) [reply bson/bsoncodec/struct_codec.go]: You're right, but I have no idea how to RegisterEncoder for a nil value in the current registry. Maybe we should modify these encoders to add pointer support?

**Comment by skriptble** [comment]
Hi @pandada8,

We've found a different approach that requires less code overall and makes the code easier to maintain, so I'm going to close this pull request. Thank you for all your hard work on this pull request, and I'm sorry that we cannot incorporate these changes into the BSON library at this time. Please do contribute in the future! We always welcome contributions from the community.

If you want to follow along with the progress of the new design, you can do so [here](https://review.gerrithub.io/c/mongodb/mongo-go-driver/+/435374).

Thanks again for your hard work!
--Kris
